File: Behavioural-Patterns/Memento/C#/Memento/Implementations/History.cs
public class History {
    private Stack<Canvas.CanvasState> undoStack = new Stack<Canvas.CanvasState>();
    
    public void save(Canvas canvas){
        undoStack.Push(canvas.createMemento());
        Console.WriteLine("Saved canvas state to history.");
    }

    public void undo(Canvas canvas){
        if(undoStack.Count == 0){
            Console.WriteLine("Nothing to undo.");
            return;
        }
        Canvas.CanvasState memento = undoStack.Pop();
        canvas.restore(memento);
    }
}

File: Behavioural-Patterns/Memento/Java/Implementations/History.java
package Implementations;

import java.util.Stack;

import Implementations.Canvas.CanvasState;

public class History {
    private Stack<CanvasState> undoStack = new Stack<>();
    
    public void save(Canvas canvas){
        undoStack.add(canvas.createMemento());
        System.out.println("Saved canvas state to history.");
    }

    public void undo(Canvas canvas){
        if(undoStack.isEmpty()){
            System.out.println("Nothing to undo.");
            return;
        }
        CanvasState memento = undoStack.pop();
        canvas.restore(memento);
    }
}
